File: selenide_testng_simple_with_wdm/src/test/java/pages/CatalogPage.java
package pages;

import com.codeborne.selenide.CollectionCondition;
import com.codeborne.selenide.ElementsCollection;
import com.codeborne.selenide.SelenideElement;

import static com.codeborne.selenide.Condition.*;
import static com.codeborne.selenide.Selenide.$;
import static com.codeborne.selenide.Selenide.$$;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.Is.is;

public class CatalogPage {

    SelenideElement catalogTitle = $(".title");
    SelenideElement cartIcon = $("#shopping_cart_container");
    SelenideElement menuIcon = $("#react-burger-menu-btn");
    SelenideElement dropdownSortingMenu = $(".product_sort_container");
    SelenideElement inventoryTableContainer = $("#inventory_container");
    ElementsCollection inventoryItems = $$(".inventory_item");
    SelenideElement cartProductBadge = $(".shopping_cart_badge");

    public void validateMainCatalogPageElementsVisible(){
        catalogTitle.shouldBe(visible);
        cartIcon.shouldBe(visible);
        menuIcon.shouldBe(visible);
        dropdownSortingMenu.shouldBe(visible);
        inventoryTableContainer.shouldBe(visible);
        for (SelenideElement e : inventoryItems){
            e.shouldBe(visible);
        }
    }

    public void validateCatalogIsNotEmpty(){
        inventoryItems.shouldBe(CollectionCondition.sizeGreaterThan(0));
    }

    public void addItemToCart() {
        inventoryItems.get(0).$("button").click();
    }

    public void cartShouldHaveItem(){
        var itemAddRemoveBtnText = inventoryItems.get(0).$("button").getText();
        assertThat(itemAddRemoveBtnText,is("Remove"));

        var numberOfItemsInCart = cartProductBadge.getText();
        assertThat(numberOfItemsInCart,is("1"));
    }

    public void selectSortByText(String text){
        dropdownSortingMenu.selectOption(text);
    }

    public void selectWorstPriceSort(){
        this.selectSortByText("Price (high to low)");
    }

    public void validateItemPrice(int itemPosition, String price){
        inventoryItems.get(itemPosition - 1).$(".inventory_item_price").shouldHave(text(price));
    }

    public void validateWorstPriceIsFirstItem(){
        this.validateItemPrice(1, "$49.99");
    }
}

File: cypress_12_with_gherkin/cypress/pages/catalog.page.js
//.title

/// <reference types="cypress" />

export class CatalogPage {

    locators = {
        catalogTitle: () => cy.get('.title'),
        cartIcon: () => cy.get('#shopping_cart_container'),
        menuIcon: () => cy.get('#react-burger-menu-btn'),
        sortDropdown: () => cy.get('.product_sort_container'),
        inventoryTableContainer: () => cy.get('#inventory_container'),
        inventoryItems: () => cy.get('.inventory_item'),
        productBackpack: () => cy.get('#add-to-cart-sauce-labs-backpack'),
        cartProductsBadge: () => cy.get('.shopping_cart_badge')
    }

    catalogPageVisible(){
        this.locators.catalogTitle().should('be.visible')
        this.locators.cartIcon().should('be.visible')
        this.locators.menuIcon().should('be.visible')
        this.locators.sortDropdown().should('be.visible')
        this.locators.inventoryTableContainer().should('be.visible')
    }

    shouldHaveInventoryItems(){
        this.locators.inventoryItems().then(els =>{
            expect(els.length).greaterThan(0)
        })
    }

    addProduct(product){
        if(product.toUpperCase() === "BACKPACK"){
            this.locators.productBackpack().click()
        }else{
            throw Error("Product not available for testing")
        }
    }

    cartProductsNrIs(number){
        this.locators.cartProductsBadge().should('have.text', number)
    }

}